### PR TITLE
Fixing fish hook exploits, bait not being consumed, race conditions and achievement

### LINKED ISSUE #73014: fishing rods can be used to haunt items
initially reported to me by Stocs: 
image

Step 1. Cast a fishing line on an item.
Step 2. Without moving far enough away to break the line, pick up the item.
Step 3. Put the item in someone else's inventory.
Step 4. You can now use the fishing rod to pull the item one tile closer to you from any range.

The REAL sauce here is that pulling the item closer to you while it's in someone else's inventory results in the item being in two places at once (both on the floor and in the other person's inventory), leading to some odd effects.

https://youtu.be/rebromHJqAM

### LINKED ISSUE #77265: You can (kind of) duplicate items with fishing rods
Reporting client version: 514.1589

## Round ID:
[211457](https://scrubby.melonmesa.com/round/211457)

## Testmerges:
- [Bumps rust-g to `3.0.0` ](https://www.github.com/tgstation/tgstation/pull/76663)

## Reproduction:

1) With a fishing rod, click on any item.
2) Have someone else pick up the item
3) Use the fishing rod (press Z)

This duplicates the item in kind of a weird quantum state. Does really weird shit with guns, in that if the person with the rod drops the copy they made and the person who has the original fires, the bullets come out of the duplicated gun.

### LINKED ISSUE #78187: Legendary fisher achievement is broken
## Reproduction:

>fish
>get 0 fish cos minigame sucks
>grind legendary skill anyway
>cool hat but no achievement 
Screenshot_114
Either only one person per round can get this achievement or admins are just built different.
Screenshot_115

### LINKED ISSUE #78204: the batshit insanity that is quantum fishing rods
## Round ID:

214191

## Testmerges:
#78124 

## Reproduction:

Get a fishing rod. Get a second player. fish the item. make the second player pick it up. drag it out of their hands by trying to fish it to you (yet its still in their hands). And its only the beginning. If you were to interact with an mmi this way, it would speak from 2 places: on second player's icon and on the floor (or on you if you picked it up). in fact, 3 people can have the same thing. And if you snatch someone's poly belt while they are polymorphed and pick it up, they will instantly turn into themselves. Maybe there are more insanities, but im unsure

### LINKED ISSUE #78210: Fishing with the rod from the Settler Quirk has become a Sisyphean task
## Reproduction: 
I was using the rod and line given by the Settler quirk, so a regular rod, line, and basic hook. Bait used was a dough ball. I successfully completed the minigame at least a dozen times, and caught nothing. It did not say the fish had gotten away, or that something had failed, it simply said 'stopped fishing' in the balloon alert. The bait was not used up, but my fishing skill did increase. I think I got all the way to expert before giving up because I wasnt catching anything. I dont know if this is a bug with the quirk items or the portal itself.
Extra details that may or may not be relevant?: This was on northstar, and the food glitch was also active at this time. [That issue was report here](https://github.com/tgstation/tgstation/issues/78192)

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
The bait wasn't being consumed properly because it was badly coded. Whoopsies.
The achievement would be unlocked upon gaining legendary rank ONLY IF you won the minigame. My bad.
Upon win (and loss), the fishing minigame would often be prematurely closed before ui act could run due to internal calls to do that. A classic example of race condition.
Last but not least, the fishing hook exploit that would allow items to be in multiple places, including nullspace, and cause a load of issues.

## Why It's Good For The Game
This PR should fix all of them. Closes #78210, closes #78204, closes #78187, closes #77265, closes #77265, closes #73014.

## Changelog

:cl:
fix: Fixed a race condition that made fishing yield no reward way too often.
fix: The legendary fisher achievement is awarded even if you don't win the minigame.
fix: Fixed a fish hook exploit.
fix: Baits are now properly consumed by caught fish and (alive) mobs.
/:cl:

### CODE REVIEW

File: code/modules/fishing/fishing_rod.dm [hunk 1]
```diff
@@ -118,11 +118,14 @@
 	return hook?.reason_we_cant_fish(target_fish_source)
 
 /obj/item/fishing_rod/proc/consume_bait(atom/movable/reward)
-	if(!reward)
+	// catching things that aren't fish or alive mobs doesn't consume baits.
+	if(!reward || !bait)
```
Comment (Jacquerel) [inline code/modules/fishing/fishing_rod.dm:122]:
```suggestion
	if(isnull(reward) || isnull(bait))
```
I believe this is marginally faster (unimportant) but more importantly it's clearer

**Review by Jacquerel — CHANGES_REQUESTED** [review]

**Review by Jacquerel — APPROVED** [review]

**Comment by leaKsi** [comment]
I don't think this fixes the fishing hook exploit from #78204, #73014, or #77265.

The fishing line now just disappears when you move (it probably shouldn't). But this doesn't stop you from performing the exploit.
It's caused by `currently_hooked_item` not being cleared properly.

**Comment by Ghommie** [comment]
> I don't think this fixes the fishing hook exploit from #78204, #73014, or #77265.

Turns out you're right. Thanks for telling me! I've fixed and moved the check for it to the proc that actually sends the line snapped signal that clears the hooked item, so the issue should be actually fixed now.

**Comment by Ghommie** [comment]
wait no no no no. I fucked this up again.

**Pull request merged**